neuraxle/pipeline.py

```diff
@@ -273,92 +120,153 @@ def fit(self, data_inputs, expected_outputs=None) -> 'Pipeline':
         """
         Transform then rehash ids with hyperparams and transformed data inputs
 
         :param data_container: data container to transform
+        :param context: execution context
         :return: tuple(fitted pipeline, transformed data container)
         """
-        data_container = self._transform_core(data_container)
+        data_container = self._transform_core(data_container, context)
 
         ids = self.hash(data_container.current_ids, self.hyperparams, data_container.data_inputs)
         data_container.set_current_ids(ids)
 
         return data_container
 
-    def _fit_transform_core(self, data_container) -> ('Pipeline', DataContainer):
+    def _fit_core(self, data_container: DataContainer, context: ExecutionContext) -> ('Pipeline', DataContainer):
+        """
+        After loading the last checkpoint, fit transform each pipeline steps,
+        but only fit the last pipeline step.
+
+        :param data_container: the data container to fit transform on
+        :param context: execution context
+        :return: tuple(pipeline, data_container)
+        """
+        steps_left_to_do, data_container = self._load_checkpoint(data_container, context)
+        self.setup(context)
+
+        index_last_step = len(steps_left_to_do) - 1
+
+        new_steps_as_tuple: NamedTupleList = []
+
+        for index, (step_name, step) in enumerate(steps_left_to_do):
+            sub_step_context = context.push(step)
+            step.setup(sub_step_context)
+
+            if index != index_last_step:
+                step, data_container = step.handle_fit_transform(data_container, sub_step_context)
+            else:
+                step, data_container = step.handle_fit(data_container, sub_step_context)
+
+            new_steps_as_tuple.append((step_name, step))
+
+        self.steps_as_tuple = self.steps_as_tuple[
+                              :len(self.steps_as_tuple) - len(steps_left_to_do)] + new_steps_as_tuple
+
+        return self, data_container
+
+    def _fit_transform_core(self, data_container: DataContainer, context: ExecutionContext) -> (
+            'Pipeline', DataContainer):
         """
         After loading the last checkpoint, fit transform each pipeline steps
 
         :param data_container: the data container to fit transform on
+        :param context: execution context
         :return: tuple(pipeline, data_container)
         """
-        steps_left_to_do, data_container = self._load_checkpoint(data_container)
+        steps_left_to_do, data_container = self._load_checkpoint(data_container, context)
+        self.setup(context)
 
         new_steps_as_tuple: NamedTupleList = []
 
         for step_name, step in steps_left_to_do:
-            step.set_parent(self)
-            step, data_container = step.handle_fit_transform(data_container)
+            sub_step_context = context.push(step)
+            step.setup(sub_step_context)
```

Comment:
I question this again despite I think the way it is coded right now is already probably the best one: should we just instead do this in one line here such as `step.setup(sub_step_context)`, and then let the step push itself to the context inside its own setup method, such that it then does `context = context.push(self)` ? 

Note: you'll probably resolve my current comment right now without further discussion. I'm just wondering.

Reply:
Same comment for other core methods and all the `setup` methods.